src/partition/repart.c

```diff
@@ -3913,7 +3914,52 @@ static int acquire_root_devno(const char *p, int mode, char **ret, int *ret_fd)
         return 0;
 }
 
+static int find_os_prefix(char **ret) {
+        char *p;
+        int r;
+
+        /* Searches the right place to look for the OS root. This is relevant in the initrd: in the initrd
+         * the host OS is typically mounted to /sysroot/ — except in setups where /usr/ is a separate
+         * partition, in which case it is mounted to /sysusr/usr/ before being moved to /sysroot/usr/. */
+
+        if (!in_initrd()) {
+                *ret = NULL; /* no prefix */
+                return 0;
+        }
+
+        r = path_is_mount_point("/sysroot", NULL, 0);
+        if (r < 0 && r != -ENOENT)
+                log_debug_errno(r, "Failed to determine whether /sysroot/ is a mount point, assuming it is not: %m");
+        else if (r > 0) {
+                log_debug("/sysroot/ is a mount point, assuming it's the prefix.");
+
+                p = strdup("/sysroot");
+                if (!p)
+                        return -ENOMEM;
+
+                *ret = p;
+                return 0;
+        }
+
+        r = path_is_mount_point("/sysusr/usr", NULL, 0);
+        if (r < 0 && r != -ENOENT)
+                log_debug_errno(r, "Failed to determine whether /sysusr/usr is a mount point, assuming it is not: %m");
+        else if (r > 0) {
+                log_debug("/sysusr/usr/ is a mount point, assuming /sysusr/ is the prefix.");
+
+                p = strdup("/sysusr");
+                if (!p)
+                        return -ENOMEM;
+
+                *ret = p;
+                return 0;
+        }
```

Comment:
Hmm, shouldn't the order be reversed? Let's say we have already mounted `/sysroot`, but haven't gotten around to `/sysroot/usr` yet. We'll return a prefix where there is no data, even though it might be available under `/sysusr/usr`.

Also, returning -ENOENT seems risky... Why not just return `/sysroot`, to avoid a failure mode? It's not nice for the tool to fail if stuff is not mounted fully yet.

Reply:
true